art/estimators/object_detection/pytorch_detection_transformer.py

```diff
@@ -0,0 +1,998 @@
+
+    return iou - (area - union) / area
+
+
+class PyTorchDetectionTransformer(ObjectDetectorMixin, PyTorchEstimator):
+    """
+    This class implements a model-specific object detector using DEtection TRansformer (DETR)
+    and PyTorch following the input and output formats of torchvision.
+    """
+
+    MIN_IMAGE_SIZE = 800
+    MAX_IMAGE_SIZE = 1333
+    estimator_params = PyTorchEstimator.estimator_params + ["attack_losses"]
+
+    def __init__(
+        self,
+        model: "torch.nn.Module" = None,
+        input_shape: Tuple[int, ...] = (3, 800, 800),
+        clip_values: Optional["CLIP_VALUES_TYPE"] = None,
+        channels_first: Optional[bool] = True,
+        preprocessing_defences: Union["Preprocessor", List["Preprocessor"], None] = None,
+        postprocessing_defences: Union["Postprocessor", List["Postprocessor"], None] = None,
+        preprocessing: "PREPROCESSING_TYPE" = None,
+        attack_losses: Tuple[str, ...] = (
+            "loss_ce",
+            "loss_bbox",
+            "loss_giou",
+        ),
+        device_type: str = "gpu",
+    ):
+        """
+        Initialization.
+
+        :param model: DETR model. The output of the model is `List[Dict[Tensor]]`, one for each input image. The
+                      fields of the Dict are as follows:
+
+                      - boxes (FloatTensor[N, 4]): the predicted boxes in [x1, y1, x2, y2] format, with values \
+                        between 0 and H and 0 and W
+                      - labels (Tensor[N]): the predicted labels for each image
+                      - scores (Tensor[N]): the scores or each prediction
+        :param input_shape: Tuple of the form `(height, width)` of ints representing input image height and width
+        :param clip_values: Tuple of the form `(min, max)` of floats or `np.ndarray` representing the minimum and
+               maximum values allowed for features. If floats are provided, these will be used as the range of all
+               features. If arrays are provided, each value will be considered the bound for a feature, thus
+               the shape of clip values needs to match the total number of features.
+        :param channels_first: Set channels first or last.
+        :param preprocessing_defences: Preprocessing defence(s) to be applied by the estimator.
+        :param postprocessing_defences: Postprocessing defence(s) to be applied by the estimator.
+        :param preprocessing: Tuple of the form `(subtrahend, divisor)` of floats or `np.ndarray` of values to be
+               used for data preprocessing. The first value will be subtracted from the input. The input will then
+               be divided by the second one.
+        :param device_type: Type of device to be used for model and tensors, if `cpu` run on CPU, if `gpu` run on GPU
+                            if available otherwise run on CPU.
+        """
+        import torch
+
+        class HungarianMatcher(torch.nn.Module):
```

Comment:
We need to mention the license of these functions in their docstrings.

Reply:
Added reference to Apache 2.0